nsenter/utils.go

```diff
@@ -17,21 +17,103 @@
 package nsenter
 
 import (
+	"os"
+	"strings"
+
 	"golang.org/x/sys/unix"
 )
 
-func processPayloadMounts(mountSysfs, mountProcfs bool) error {
-	var flags uintptr = unix.MS_NOSUID | unix.MS_NOEXEC | unix.MS_NODEV | unix.MS_RELATIME
+type payloadMountsInfo struct {
+	sysfsMountpoint  string
+	procfsMountpoint string
+	cleanup          func()
+}
+
+// This function runs when the sysbox-fs nsenter helper process requests to
+// mount procfs or sysfs.
+//
+// Note that the container process should do so from an unshared mount
+// namespace, such that the mounts are NOT visible inside the container
+// (otherwise container processes will see the nsenter process mounts which are
+// inherited from sysbox-fs mounts, thus leaking host info into the container).
+//
+// Note also that the nsenter process mounts the real procfs and sysfs, not the
+// sysbox-fs emulated ones. That's because the nsenter process is not under
+// seccomp-notify intercepts on mount syscalls as the container processes are.
+func processPayloadMounts(mountSysfs, mountProcfs bool) (*payloadMountsInfo, error) {
+	var (
+		flags            uintptr
+		sysfsMountpoint  string
+		procfsMountpoint string
+		err              error
+	)
+
+	flags = unix.MS_NOSUID | unix.MS_NOEXEC | unix.MS_NODEV | unix.MS_RELATIME
+
+	// Ideally we want to mount procfs on /proc and sysfs on /sys, inside the
+	// container; and since the mounts are done by the nsenter process in a
+	// dedicated mount-ns, the container processes won't see them. While it's
+	// possible for the nsenter process to mount procfs on top of the container's
+	// /proc, turns out it's not possible to mount sysfs on top of the
+	// container's /sys (the kernel returns a "resource busy" error). Thus, we
+	// mount sysfs on a temporary ephemeral dir inside the container, at
+	// /.sysbox-sysfs-<random-id>. Note that only the nsenter process can see the
+	// mount (because it operates in a dedicated mount-ns). The container
+	// processes will never see the mount, and therefore the
+	// /.sysbox-sysfs-<random-id> dir will always look empty to container
+	// processes.
 
 	if mountSysfs {
-		if err := unix.Mount("sysfs", "/sys", "sysfs", flags, ""); err != nil {
-			return err
+		sysfsMountpoint, err = os.MkdirTemp("/", ".sysbox-sysfs-")
+		if err != nil {
+			return nil, err
+		}
+		if err = unix.Mount("sysfs", sysfsMountpoint, "sysfs", flags, ""); err != nil {
+			os.RemoveAll(sysfsMountpoint)
+			return nil, err
+		}
+	}
+	cleanupSysfs := func() {
```

Comment:
Shouldn't these cleanup routines be within the `if mountSyfs` and `if mountProcfs` code blocks? Why do you want to return a cleanup() function when no `mount()` has been made?

Reply:
Yes exactly, the cleanups only work when sysfs or procfs are mounted.

The reason I needed them is because the cleanup (e.g., removal of the sysfs mountpoint) needs to occur after the event handling (done by the caller of procssPayloadMounts) has taken place.